[WIP] PHP: Make the default package more sane

etu opened the pull request:
###### Motivation for this change
Our old PHP derivation was crazy big with a crazy amount of enabled extensions that also had a crazy amount of defaults that we pulled in, yet excluded some extensions that would have been nice to have by default.

This is a follow up on: https://github.com/NixOS/nixpkgs/pull/82348 to utilize https://github.com/NixOS/nixpkgs/pull/79377 more.

I made a chart of extensions in org-mode: http://ix.io/2evY

We had several insane defaults such as `ldap`, `imap`, `pdo_odbc`, `gd` but still missed things that are quite sane such as `opcache`.

#### The current PHP
The current closure size is:
```
245697960    /nix/store/4ga497cqvz04vcyw9lf7s74ymrc6r28g-php-7.4.3
```

And this contains things (among very others) like:
```
56161288    /nix/store/2zsgfrkp0hn878rnrp0azpmy76ljm2b0-linux-pam-1.3.1
56167840    /nix/store/y5166rf2lkxkf3991qm390vn92czgfgh-openldap-2.4.49
```

##### The new smaller PHP
Here we have a much smaller closure size:
```
141377496    /nix/store/kij6n7rxzj9y54fg8ik44hhii3s824yr-php-7.4.3
```

##### Breaking changes
I've still have to track down packages in nixpkgs that will break from this, some examples are probably nextcloud among others due to the missing database extensions.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

etu committed:
php: Make a slimmer php as default

etu committed:
phpPackages.exts: Drop uneeded extensions

etu committed:
php: Add release log entry for the php changes

jtojnar commented:
Only checked the change list and it looks okay to me except for mbstring. That is something I would expect enabled by default these days.

Ma27 commented on pkgs/development/interpreters/php/default.nix:
```diff
@@ -235,6 +148,12 @@ let
 
+    patches = [ ./fix-paths-php7.patch ] ++ extraPatches;
+
+    stripDebugList = "bin sbin lib modules";
```
Any reason why we don't move debugging-symbols into a separate `debug`-output (is usually doable with `separateDebugInfo = true`)?

etu replied:
This isn't actually any change except that I moved it around in the file to make it look better (in my opinion).

May be doable, haven't tried.

etu replied:
Seems to work fine, fixed in 0d392490483

Ma27 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -86,6 +86,53 @@
+             pdo_pgsql
+           ]; })
+         ];
+       </programlisting>
```
Would it make sense to provide e.g. a `phpMinimal` or a `phpExtended` (with opinionated extensions) or even a `phpFull`?

etu replied:
I would say that the one I provide in this PR is a good enough base, and if you want to build a minimal one you can do that. And a full one would be to add all the extensions to get a full one. And that shouldn't require any compiling so it's fast to build.

And I find the one I provide by default in this PR small enough so we don't need a minimal one in my opinion.

I was thinking about it before though, that's why the minimal flag is there at all :)

But the minimal one has some issues to build it completely minimal. That is if you have a minimal build without `--enable-mysqlnd` you're not able to build `phpPackages.exts.pdo_mysql` because the php used that provides `phpize` needs to be built with `--enable-mysqlnd` to have `mysqlnd.h` available.

This problem could be solved to use a less minimal php to build the extensions and then use a minimal as base for the package used. But then you would end up with both of those in the closure so I'd rather avoid that... :-)

Ma27 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -86,6 +86,53 @@
+       additional extensions enabled. See following snippet to install a PHP with <literal>gd</literal>,
+       <literal>pdo_mysql</literal> and <literal>pdo_pgsql</literal> added:
+
+       <programlisting>
```
Is this properly indented in the rendered manual? IIRC `<programlisting>`-tags are completely unindented in our docs.

etu replied:
I wouldn't know, I'm not very familiar with docbook. I was reading previous releasenotes to find good examples. And I think it is based on the firejail entry here: https://nixos.org/nixos/manual/release-notes.html#sec-release-18.09-highlights

Ma27 replied:
Also no idea :D

I'd recommend to review it using e.g. `firefox ./result/share/doc/nixos/index.html` after running `nix build --print-build-logs -f nixos/release.nix --arg supportedSystems '[ builtins.currentSystem ]' --option builders '' manual` (I really command an alias for the building stuff at least :p)

etu replied:
It does indent, but I've fixed up the indentation to align to the left in the rendered manual :)
![image](https://user-images.githubusercontent.com/461970/76894114-ce5b7880-688d-11ea-9612-2f822e963e70.png)

jtojnar replied:
You can just use `nix-build nixos/release.nix -A manualHTML.x86_64-linux`. And `nix-build pkgs/top-level/release.nix -A manual` for nixpkgs.

etu commented on pkgs/development/interpreters/php/default.nix:
```diff
@@ -1,196 +1,111 @@
-  , phpdbgSupport ? config.php.phpdbg or true
+
+  # Build a minimal php
+  , minimalBuild ? config.php.minimal or false
```
We don't seem to have any doc at all for PHP... I looked into the language-frameworks directory and found python there for instance... And yeah. Nobody has even started a php document...

jtojnar replied:
Nixpkgs docs are supposed to be for building packages (package build support). I guess section for PHP build support like `composer2nix` might make sense but user docs is better suited for NixOS docs.

etu replied:
`composer2nix` isn't in nixpkgs at all (:disappointed:) and isn't documented either because of that.

It would be great to document the `php.buildEnv` behavior somewhere other than the code and in pull requests.

etu commented:
I have just completed the `nixpkgs-review` run again. All packages that aren't marked as broken does build. I will assume that all packages within `phpPackages` does work just as is.

The ones I'm not sure if they work during runtime are the following:
 - [ ] adminer
 - [ ] arcanist
 - [ ] drush
 - [ ] icingaweb2
 - [ ] kcachegrind
 - [ ] lsp-plugins
 - [ ] matomo-beta
 - [ ] nagios
 - [ ] nextcloud-news-updater
 - [ ] phoronix-test-suite
 - [ ] piwik
 - [ ] pulseeffects
 - [ ] qcachegrind
 - [ ] unit
 - [ ] wp-cli
 - [ ] yle-dl

jtojnar commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -86,6 +86,53 @@
+         ];
+       </programlisting>
+
+       All native PHP extensions are available under <literal><![CDATA[phpPackages.exts.<name?>]]></literal>.
```
There is not documentation for packages except for the package list and that is not linkable. Docbook has `<package>` element though.

etu replied:
I've switched to the `<package>` element and mentioned that the listed extensions that aren't loaded as default are available as loadable extensions. That makes it at least a bit discoverable to the end user reading the release log. See 42b24cf064e

etu commented on pkgs/development/interpreters/php/default.nix:
```diff
@@ -1,196 +1,111 @@
-  , readlineSupport ? config.php.readline or true
-  , sqliteSupport ? config.php.sqlite or true
-  , soapSupport ? (config.php.soap or true) && (libxml2Support)
-  , zlibSupport ? config.php.zlib or true
```
I've added these to the release log now in 87616d92cb5 :)

etu committed:
php-embed: Drop the declaration of the php-embed attributes

etu committed:
php-unit: Drop the declaration of the php-unit attributes since they aren't used

etu committed:
doc/php: Add initial documentation for PHP

flokli commented:
Recap: I had a very nice (remote) session with @etu hacking on this today.

We need to do some more thinking about how to make this a bit less obstrusive (it currently breaks every NixOS module shipping software requiring some sort of database access).

This could be solved by explicitly listing the required modules. However, the current approach istn't that nice. We could improve it by making the "default" php package configurable through system configuration, or by adding (some few) extensions.

There were also some cleanups (`php-unit` and `php-embed`), and the initial addition of some docs about the PHP package tooling.

etu committed:
php: init phpXYbase packages and make the default php extended by default

etu committed:
phpPackages.exts.gmp: Enable on darwin

etu commented:
@GrahamcOfBorg php74Packages.exts.gmp php74 php73 php72 php74base php73base php72base

etu commented:
@GrahamcOfBorg php74Packages.exts.gmp php74 php73 php72 php74base php73base php72base

etu commented:
@GrahamcOfBorg php74Packages.exts.gmp php74 php73 php72 php74base php73base php72base

etu committed:
unit: Make unit use phpbase packages

etu commented:
@GrahamcOfBorg php74Packages.exts.gmp php74 php73 php72 php74base php73base php72base

talyz committed:
php: Build an even slimmer base

This moves yet more extensions from the base build to
phpPackages.ext. Some of the extensions are a bit quirky and need
patching for this to work, most notably mysqlnd.

Two new parameters are introduced for mkExtension - internalDeps and
postPhpize. internalDeps is used to specify which other internal
extensions the current extension depends on, in order to provide them
at build time. postPhpize is for when patches and quirks need to be
applied after running phpize.

talyz committed:
phpPackages.exts: Run tests by default

The tests for many of the extensions run just fine, for some a small
portion fail. This runs the tests by default and disables the tests
extensions with any failing tests.

talyz committed:
php: Make buildEnv recursive + take extension deps into account

A slight rewrite of buildEnv which:

1. Makes buildEnv recursively add itself to its output, so that it can
   be accessed from any php derivation.

2. Orders the extension text strings according to their internalDeps
   attribute - dependencies have to be put before dependants in the
   php.ini or they will fail to load due to missing symbols.

talyz committed:
php: Add opcache to default extensions

Opcache significantly increases PHP's performance and is by default
enabled on Debian based distributions. Not having it enabled by
default results in a puzzling performance loss for anyone attempting
to migrate from Debian/Ubuntu to NixOS who is unaware of
this. Therefore, enable it by default.

etu committed:
Merge branch 'talyz-slim-down-default-php' into slim-down-default-php

etu committed:
php: Drop phpbase packages since they aren't needed anymore

etu committed:
Merge branch 'master' into slim-down-default-php

etu committed:
docs/rl: Update release log to match the updates to the default attributes

etu committed:
php: Re-introduce phpbase for builds of packages to reduce complexity

etu committed:
phpPackages.exts: Enable more extensions to build with PHP less than 7.4

etu committed:
php: Add json as default extension

talyz committed:
phpPackages.exts.mysqlnd: Fix building with PHP <7.4

etu committed:
Merge branch 'talyz-slim-down-default-php2' into slim-down-default-php

talyz committed:
phpPackages.exts.opcache: Fix build for PHP <7.4

Older versions of PHP have a bug where header files are included in
the wrong order. This is only an issue when compiling the opcache
extension as a separate module, like we do.

talyz committed:
php: Make mkExtension pass on additional args to mkDerivation

talyz committed:
phpPackages.exts.mysqlnd: Use a proper patch instead of sed

Also, document why the patch is needed.

etu committed:
Merge branch 'talyz-slim-down-default-php3' into slim-down-default-php

etu commented:
This PR has grown, not because of bad reasons. And things have gone quite back and forth a few times and I don't like that. But I still want to do this. So I made #83514 where I've cherry-picked out the things we need and resolved all the conflicts and made the history cleaner and nicer.
